components/mediators/cache/org.wso2.carbon.mediator.cache/src/main/java/org/wso2/carbon/mediator/cache/CacheMediator.java

```diff
@@ -0,0 +1,693 @@
+	 * up for a hit, if found; then the specified named or anonymous sequence is executed or marks
+	 * this message as a response and sends back directly to client.
+	 *
+	 * @param synCtx incoming request message
+	 * @param synLog the Synapse log to use
+	 * @return should this mediator terminate further processing?
+	 * @throws ClusteringFault if there is an error in replicating the cfgCtx
+	 */
+	private boolean processRequestMessage(MessageContext synCtx,
+	                                      SynapseLog synLog) throws ClusteringFault {
+
+		if (collector) {
+			handleException("Request messages cannot be handled in a collector cache", synCtx);
+		}
+
+		OperationContext opCtx = ((Axis2MessageContext) synCtx).getAxis2MessageContext().getOperationContext();
+		String requestHash = null;
+
+		try {
+			requestHash = digestGenerator.getDigest(((Axis2MessageContext) synCtx).getAxis2MessageContext());
+			synCtx.setProperty(CachingConstants.REQUEST_HASH, requestHash);
+		} catch (CachingException e) {
+			handleException("Error in calculating the hash value of the request", e, synCtx);
+		}
+
+		if (synLog.isTraceOrDebugEnabled()) {
+			synLog.traceOrDebug("Generated request hash : " + requestHash);
+		}
+
+		RequestHash hash = new RequestHash(requestHash);
+		CachableResponse cachedResponse = getMediatorCache().get(requestHash);
+		org.apache.axis2.context.MessageContext msgCtx = ((Axis2MessageContext) synCtx).getAxis2MessageContext();
+		opCtx.setProperty(CachingConstants.REQUEST_HASH, requestHash);
+
+		byte[] responseEnvelop;
+		Map<String, Object> headerProperties;
+		if (cachedResponse != null && (responseEnvelop = cachedResponse.getResponseEnvelope()) != null) {
+			// get the response from the cache and attach to the context and change the
+			// direction of the message
+			if (!cachedResponse.isExpired()) {
+				if (synLog.isTraceOrDebugEnabled()) {
+					synLog.traceOrDebug("Cache-hit for message ID : " + synCtx.getMessageID());
+				}
+				cachedResponse.setInUse(true);
+				// mark as a response and replace envelope from cache
+				synCtx.setResponse(true);
+				opCtx.setProperty(CachingConstants.CACHED_OBJECT, cachedResponse);
+
+				SOAPEnvelope omSOAPEnv = null;
+
+				try {
+					if (msgCtx.isDoingREST()) {
+						if ((headerProperties = cachedResponse.getHeaderProperties()) != null) {
+
+							omSOAPEnv = SOAPMessageHelper
+									.buildSOAPEnvelopeFromBytes(responseEnvelop, cachedResponse.isSOAP11());
+							msgCtx.removeProperty("NO_ENTITY_BODY");
+							msgCtx.removeProperty(Constants.Configuration.CONTENT_TYPE);
+							msgCtx.setProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS,
+							                   headerProperties
+									                   .get(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS));
+							msgCtx.setProperty(Constants.Configuration.MESSAGE_TYPE,
+							                   headerProperties.get(Constants.Configuration.MESSAGE_TYPE));
+						}
```

Comment:
Thanks a lot for pointing this out. Issue is fixed with PR https://github.com/wso2/carbon-mediation/pull/452